File: labs/Lab3/scripts/pid_control.py
#!/usr/bin/env python
from __future__ import print_function
import sys
import math
import numpy as np

#ROS Imports
import rospy
from sensor_msgs.msg import Image, LaserScan
from ackermann_msgs.msg import AckermannDriveStamped, AckermannDrive

#PID CONTROL PARAMS
kp = 0.55
kd = 0.2
ki = 0.0001
servo_offset = 0.0
prev_error = 0.0 
error = 0.0
integral = 0.0

#WALL FOLLOW PARAMS
ANGLE_RANGE = 360 
DESIRED_DISTANCE_RIGHT = 0.9 
DESIRED_DISTANCE_LEFT = 0.55
VELOCITY = 2 # meters per second
CAR_LENGTH = 0.5 # Traxxas Rally is 20 inches or 0.5 meters

class WallFollow:
    """ Implement Wall Follow on the car
    """
    def __init__(self):
        #Topics & Subs, Pubs
        lidarscan_topic = '/scan'
        drive_topic = '/drive'
        self.lidar_sub = rospy.Subscriber(lidarscan_topic, LaserScan, self.lidar_callback)
        self.drive_pub = rospy.Publisher(drive_topic, AckermannDriveStamped, queue_size=10)

    def getRange(self, data, angle):
        # data: single message from topic /scan
        # angle: between -45 to 225 degrees, where 0 degrees is directly to the right
        # Outputs length in meters to object with angle in lidar scan field of view
        #make sure to take care of nans etc.
	index = angle * 3
        dist = data.ranges[index]
        if np.isnan(dist) or np.isinf(dist):
            return 0.0
        else:
            return dist


    def pid_control(self, error, velocity):
        global integral
        global prev_error
        global kp
        global ki
        global kd
	angle = 0.0
        integral += error
        derivative = error - prev_error
        angle = -(kp*error + ki*integral + kd*derivative)
        prev_error = error
 	rospy.loginfo('Angle: {}'.format(math.degrees(angle)))

	if 0 <= math.degrees(abs(angle)) <= 10:
            velocity = 1.5
        elif 10 < math.degrees(abs(angle)) <= 20:
            velocity = 1
        else:
            velocity = 0.5
	
        drive_msg = AckermannDriveStamped()
        drive_msg.header.stamp = rospy.Time.now()
        drive_msg.header.frame_id = "laser"
        drive_msg.drive.steering_angle = angle
        drive_msg.drive.speed = velocity
        self.drive_pub.publish(drive_msg)

    def followLeft(self, data, leftDist):
        #Follow left wall as per the algorithm
        a = self.getRange(data,210)
        b = self.getRange(data,270)
        alpha = math.atan((a * math.cos(math.radians(60)) - b) / (a * math.sin(math.radians(60))))
	rospy.loginfo('Alpha: {}'.format(alpha))
        D_t = b*math.cos(alpha)
        D_tplus1 = D_t + math.sin(alpha)*CAR_LENGTH
	error = (DESIRED_DISTANCE_LEFT - D_tplus1)
	rospy.loginfo('Dt: {}'.format(D_tplus1))
	rospy.loginfo('Error: {}'.format(error))
        return error

    def lidar_callback(self, data):
        error = self.followLeft(data, DESIRED_DISTANCE_LEFT)
        self.pid_control(error, VELOCITY)

def main(args):
    rospy.init_node("WallFollow_node", anonymous=True)
    wf = WallFollow()
    rospy.spin()

if __name__=='__main__':
	main(sys.argv)

File: labs/Lab3/scripts/pid_control.cpp
#include <ros/ros.h>
#include <sensor_msgs/LaserScan.h>
#include <ackermann_msgs/AckermannDriveStamped.h>
#include <cmath>
#include <algorithm>

// PID CONTROL PARAMS
double kp = 0.55;
double kd = 0.2;
double ki = 0.0001;
double servo_offset = 0.0;
double prev_error = 0.0;
double error = 0.0;
double integral = 0.0;

// WALL FOLLOW PARAMS
const int ANGLE_RANGE = 360;
const double DESIRED_DISTANCE_RIGHT = 0.9;
const double DESIRED_DISTANCE_LEFT = 0.55;
const double VELOCITY = 2; 
const double CAR_LENGTH = 0.5; 

class WallFollow
{
public:
    WallFollow()
    {
        // Topics & Subs, Pubs
        std::string lidarscan_topic = "/scan";
        std::string drive_topic = "/drive";
        lidar_sub = nh.subscribe(lidarscan_topic, 1000, &WallFollow::lidar_callback, this);
        drive_pub = nh.advertise<ackermann_msgs::AckermannDriveStamped>(drive_topic, 10);
    }

    void lidar_callback(const sensor_msgs::LaserScan::ConstPtr& data)
    {
        double error = followLeft(data, DESIRED_DISTANCE_LEFT);
        pid_control(error, VELOCITY);
    }

private:
    ros::NodeHandle nh;
    ros::Publisher drive_pub;
    ros::Subscriber lidar_sub;

    double getRange(const sensor_msgs::LaserScan::ConstPtr& data, int angle)
    {
        int index = angle * 3;
        double dist = data->ranges[index];
        if (std::isnan(dist) || std::isinf(dist)) {
            return 0.0;
        } else {
            return dist;
        }
    }

    void pid_control(double error, double velocity)
    {
        double angle = 0.0;
        integral += error;
        double derivative = error - prev_error;
        angle = -(kp * error + ki * integral + kd * derivative);
        prev_error = error;
        ROS_INFO("Angle: %f", angle * 180.0 / M_PI);

        if (std::abs(angle) <= 10.0 * M_PI / 180.0) {
            velocity = 1.5;
        } else if (std::abs(angle) <= 20.0 * M_PI / 180.0) {
            velocity = 1.0;
        } else {
            velocity = 0.5;
        }

        ackermann_msgs::AckermannDriveStamped drive_msg;
        drive_msg.header.stamp = ros::Time::now();
        drive_msg.header.frame_id = "laser";
        drive_msg.drive.steering_angle = angle;
        drive_msg.drive.speed = velocity;
        drive_pub.publish(drive_msg);
    }

    double followLeft(const sensor_msgs::LaserScan::ConstPtr& data, double leftDist)
    {
        // Follow left wall as per the algorithm
        double a = getRange(data, 210);
        double b = getRange(data, 270);
        double alpha = atan((a * cos(M_PI / 3) - b)/ (a * sin(M_PI / 3)));
        ROS_INFO("Alpha: %f", alpha);
        double D_t = b * cos(alpha);
        double D_tplus1 = D_t + sin(alpha) * CAR_LENGTH;
        double error = (DESIRED_DISTANCE_LEFT - D_tplus1);
        ROS_INFO("Dt: %f", D_tplus1);
        ROS_INFO("Error: %f", error);
        return error;
    }
};

int main(int argc, char** argv)
{
    ros::init(argc, argv, "WallFollow_node");
    WallFollow wf;
    ros::spin();
    return 0;
}

